server/service/service.go

```diff
@@ -29,6 +29,11 @@ func New(conf *config.Config) (*Service, error) {
 		return nil, fmt.Errorf("failed to create PanaceaGRPCClient")
 	}
 
+	err = panaceaClient.RegisterDataValidator(conf.PublicEndPointUrl, validatorAccount)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
IMHO, if this is done, it will be registered in panacea normally at the first run, but an error will occur during subsequent runs.
I think you should subtract this as a command.
(ex: `datavald register`)

Reply:
The `datavald register` command would be good as well.
If we wanna run the `RegisterDataValidator` automatically when the process is started, duplication checks need to be performed before executing the tx.

I'm not sure which way is better: Registration at start-up V.S. `datavald register ...`.

Reply:
You guys can decide which way would be reasonable. If Inchul is already implementing the registration at start-up with duplication checks, it's good to go with his implementation for now. If we feel the second way is better, we can change the implementation at that time. 
I guess that duplication checks are necessary for both ways.

Reply:
Yes I agree.
Let's proceed with the duplicate check first.

Reply:
So, Does I need to check the duplication of registered data validator before `RegisterDataValidator` runs?

Reply:
Yes, if you are already registered as a data validator, you can just avoid calling it.

Reply:
I have a question that I implemented duplication checks in [Panacea](https://github.com/medibloc/panacea-core/blob/9f9b3bec93d6fa2d7adf57b0d7237fe6522721c7/x/datapool/keeper/pool.go#L18). But I need to implement duplication check before `RegisterDataValidator` runs too?
Plus, if I implement duplication check before `RegisterDataValidator`, does I need to make query for data validator?

Reply:
There seem to be two ways.
First, there is a way to judge by running `RegisterDataValidator` and viewing the error response message.
The second is to query by implementing query as @inchori said.
The first method is inconvenient to always incur fees because tx has to be blown, but the code implementation is simple.
The second needs to perform a query, so we need two actions.

My opinion is the second way, but you can used  the first way it, because requirs panacea code implements.

Reply:
You may have to check if the data validator is registered or not. (I implemented a [keeper func](https://github.com/medibloc/panacea-core/blob/5f69ad06762947b77e90421a4b2487f058772fa2/x/datapool/keeper/pool.go#L72) for this, but no query for this yet)

If not, the data validator should register to Panacea first.
In general, however, if the data validator doesn't register to Panacea, any sellers can request to the validator because there is no way to get endpoint. In other words, the data validator is registered if someone request to validate their data. 

For now, the registration of data validator is just for informing their endpoint to request data validation. we can use the registration info for incentive to data validator in the future(have not thought deeply yet).

Reply:
Thanks for your comments! 
@gyuguen The second way that implementing query for registered data validator seems to be good. But how about using `datapoolkeeper.Keeper.RegisterDataValidator()` function after `panacea.NewGrpcClient(conf)`?

@hansol-medi 
Yes, exactly register data validator is now just for informing their endpoint. Also, thanks for making check duplication of registered data validator function! How about using this `datapoolkeeper.Keeper.RegisterDataValidator()` function that you implemented as I said to gyuguen.

Reply:
I've decided to use `Query` for data validator so that I can check duplicate registered data validator in Panacea. So it can be solved.

Reply:
FYI, we also need to consider having the `EditDataValidator` transaction because someone would wanna change their endpoint for some reason. Then, the duplication check login in the dataval would become complicated. For example, even if the data validator was already registered, the dataval should execute `EditDataValidator` if the registered endpoint is old.

Reply:
Yes, it needs to having `EditDataValidator` transaction too. So I will implement `Query` for data validator and after that I will implement `EditDataValidator` transaction. 
How about we change our key(dataval address) in KVstore to endpoint URL?

Reply:
I think the dataval addr should be unchangable.

Reply:
After `Query` and duplication check of data validator done, first of all I will implement `EditDataValidator` (or `UpdataDataValidator`) transaction in Panacea.
IMO, `EditDataValidator` can be sub-command of `datavald`. The data validators who want to edit themselves need to be registered data validator in Panacea.